packages/api/src/external/carequality/command/ihe-result/create-ihe-result.ts

```diff
@@ -38,12 +38,12 @@ export async function handleIHEResponse({ type, response }: IHEResult): Promise<
   let status = "failure";
 
   // Check if response is a BaseErrorResponse
-  if (isBaseErrorResponse(response)) {
-    status = "failure";
-  } else if (isDocumentQueryResponse(response) || isDocumentRetrievalResponse(response)) {
+  if (isDocumentQueryResponse(response) || isDocumentRetrievalResponse(response)) {
```

Comment:
I changed the order here. Why? Because of type narrowing! After some research:

In TypeScript, the process of refining the type of a variable within a certain scope based on conditions or type guards is known as "type narrowing". This feature helps prevent errors by ensuring that we only use properties and methods that are guaranteed to exist on the narrowed type.

In the original code, we first checked if response is a BaseErrorResponse using the isBaseErrorResponse type guard. If this condition was false, TypeScript narrowed the type of response to exclude BaseErrorResponse in the subsequent else if block. However, because BaseErrorResponse is not explicitly part of the IHEResult union, TypeScript inferred that response is of type never in the else if block. The never type in TypeScript represents a value that never occurs, which resulted in me getting an error when trying to access response.documentReference.

To resolve this, I rearranged the conditions to first check if response is a DocumentQueryResponse or DocumentRetrievalResponse. These types are part of the IHEResult union, so TypeScript can correctly infer the type of response in each block. This change ensures that we are correctly narrowing the type of response and prevents the TypeScript error.

Reply:
Yeah, always from the more specific to the more generic when checking for types.